### ISSUE TITLE
XML Markup Scheme for Marian Decoder

### ISSUE THREAD

**Issue by phikoehn** [marian-nmt/marian-dev#399]
I might be interested in taking another stab at a XML markup scheme with specified translations as in Moses:

```
The house is <x translation="klein"> small </a> .
```

This does require the decoding algorithm to have access to the attention states for each output word. This sounds rather tricky to me, since the decoder is model-agnostic. Any suggestions?

**Comment by emjotde** [comment]
Ad 2) Can you provide a compilation log? I am not aware of changes we introduced apart from changes to Warnings. If that hurts you we should take a look at that.

**Comment by snukky** [comment]
I updated the branch `xml` with the current master, and put it into the branch [xml-merge](https://github.com/marian-nmt/marian-dev/compare/xml-merge?diff=split) as the merge wasn't straightforward. The code is still far from being ready for a pull request. With the `xml` branch, I am getting segmentation faults for `beam-size 1` and random illegal memory access errors, I think, from `translator/nth_element.cu`. This happens even without `--xml-input`. The code will also need to be cleaned from debug messages and refactorized.

@phikoehn Could you provide some usage examples or test cases with expected outputs? I am not sure how to use the constrained decoding. I would like to add a few regression tests and take a look at the errors.

**Comment by phikoehn** [comment]
To use this, the flag "--xml-input" must be added. Input with specified translations is presented as

das haus ist <a translation="big"> gross </a> .

where the name of the tag (a) does not matter, but "translation" has to be used to specify a single translation. The translation has to be in byte-pair-encoded format in the target language.

**Comment by snukky** [comment]
Yes, I tried something like that (following some Moses examples), but didn't see any difference. I was decoding with the Edinburgh WMT16 EN-DE model.

**Comment by phikoehn** [comment]
Did you specify "--xml-input" when calling the decoder?

Did you never see an effect or just not always?

You should try short examples first just to be sure that it is called
correctly.

**Comment by phikoehn** [comment]
Regarding the memory access issues: check how XmlOption, XmlOptionCovered, XmlOptions, XmlOptionsList, and XmlOptionCoveredList objects are allocated. The best way would be to use smart pointers, but I could not pull that off.

**Comment by snukky** [comment]
I tried, for example:

    echo "this is <n translation="schlecht"> bad </n>" \
    | ../marian-dev/build-xml/marian-decoder -c marian.en-de.yml --xml-input

and got:

    das ist schlimm

It seems xmlSearch is executed, but I don't see any difference between this and `echo "this is bad"` w/o `--xml-input`. The n-best list scores are the same. Logs attached.

<details>
<summary>[click to see logs]</summary>
<pre><code>
[2018-12-07 22:04:10] [config] workspace: 512
[2018-12-07 22:04:10] [config] xml-alignment-weight: 1
[2018-12-07 22:04:10] [config] xml-input: true
[2018-12-07 22:04:10] [config] xml-violation-penalty: 10
...
process xml for this is <n translation=schlecht> bad </n>
vocabs_.size() = 1
called processXml
CorpusBase::addWordsToSentenceTuple
list is 0x91d0f60
batch->setXmlOptionsList(xmlOptionsList);
setXmlOptionsList 0x91d0f60
xopsl->size() = 1, 0x91c99e0
xops->size() = 0
batch->setXmlOptionsList(xmlOptionsList); OK
[2018-12-07 22:04:13] [data] Loading vocabulary from Yaml/JSON file /mnt/zisa0/romang/marian/mrt/models/wmt16_systems/en-de/vocab.de.json
pulling xmlOptionsList 0x91d0f60
xmlOptions 0x91c99e0
Hypothesis xmlOptions 0x91c99e0
still alive and kicking 4
beam=0 i=0 xml status=0/0
beam=0 i=1 xml status=0/0
beam=0 i=2 xml status=0/0
starting beam sizes 3
beam 0: 0
subbeamCount = 1
ADDING ADDITIONAL HYPOTHESES
REGULAR BEAM EXPANSION
beam 0, subbeam 0: 111
nth->setHypMask
collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 0
collectedCosts[0][0].size() = 0
build new XmlOptionCoveredList for beam 0 hyp 0 cost -0.0101005  das:12 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 1
collectedCosts[0][0].size() = 1
build new XmlOptionCoveredList for beam 0 hyp 0 cost -5.71356    dies:88 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 2
collectedCosts[0][0].size() = 2
build new XmlOptionCoveredList for beam 0 hyp 0 cost -7.04192    es:21 ...SUBBEAM 0 COST/KEY
beam 0 hyp 0>0  cost -0.0101005  das:12 ...[0]
beam 0 hyp 0>1  cost -5.71356    dies:88 ...[0]
beam 0 hyp 0>2  cost -7.04192    es:21 ...[0]
allotted: 3/3
 toBeRedistributed=0 3/3redistributed: 3/3
merge beam 0 from subbeam 0, hyp 0: das:12,-0.0101005
merge beam 0 from subbeam 0, hyp 1: dies:88,-5.71356
merge beam 0 from subbeam 0, hyp 2: es:21,-7.04192
outCosts.size() = 3
outCosts.size() = 3, localBeamSize = 3
beam 0 hyp 0>0  cost -0.0101005  das ...[0]
beam 0 hyp 0>1  cost -5.71356    dies ...[0]
beam 0 hyp 0>2  cost -7.04192    es ...[0]
beam 0 hyp 0    cost -0.0101005 [0]  das
beam 0 hyp 1    cost -5.71356   [0]  dies
beam 0 hyp 2    cost -7.04192   [0]  es
pruning the beam
histories[i]->Add(beams[i]
histories[i]->Add(beams[i] OK
remaining beam sizes 3
beam 0 hyp 0    cost -0.0101005 [0]  das
beam 0 hyp 1    cost -5.71356   [0]  dies
beam 0 hyp 2    cost -7.04192   [0]  es
DONE WITH LOOP, localBeamSize now 3
beam=0 i=0 xml status=0/0
beam=0 i=1 xml status=0/0
beam=0 i=2 xml status=0/0
starting beam sizes 3
starting beam sizes 3
beam 0: 0
subbeamCount = 1
ADDING ADDITIONAL HYPOTHESES
REGULAR BEAM EXPANSION
beam 0, subbeam 0: 111
nth->setHypMask
collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 0
collectedCosts[0][0].size() = 0
build new XmlOptionCoveredList for beam 0 hyp 0 cost -0.0170677  ist:13 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 1
collectedCosts[0][0].size() = 1
build new XmlOptionCoveredList for beam 0 hyp 1 cost -5.72473    ist:13 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 2
collectedCosts[0][0].size() = 2
build new XmlOptionCoveredList for beam 0 hyp 0 cost -6.66541    geht:171 ...SUBBEAM 0 COST/KEY
beam 0 hyp 0>0  cost -0.0170677  ist:13 ...[0]  das
beam 0 hyp 1>1  cost -5.72473    ist:13 ...[0]  dies
beam 0 hyp 0>2  cost -6.66541    geht:171 ...[0]  das
allotted: 3/3
 toBeRedistributed=0 3/3redistributed: 3/3
merge beam 0 from subbeam 0, hyp 0: ist:13,-0.0170677
merge beam 0 from subbeam 0, hyp 1: ist:13,-5.72473
merge beam 0 from subbeam 0, hyp 2: geht:171,-6.66541
outCosts.size() = 3
outCosts.size() = 3, localBeamSize = 3
beam 0 hyp 0>0  cost -0.0170677  ist ...[0]  das
beam 0 hyp 1>1  cost -5.72473    ist ...[0]  dies
beam 0 hyp 0>2  cost -6.66541    geht ...[0]  das
beam 0 hyp 0    cost -0.0170677 [0]  ist das
beam 0 hyp 1    cost -5.72473   [1]  ist dies
beam 0 hyp 2    cost -6.66541   [0]  geht das
pruning the beam
histories[i]->Add(beams[i]
histories[i]->Add(beams[i] OK
remaining beam sizes 3
beam 0 hyp 0    cost -0.0170677 [0]  ist das
beam 0 hyp 1    cost -5.72473   [1]  ist dies
beam 0 hyp 2    cost -6.66541   [0]  geht das
DONE WITH LOOP, localBeamSize now 3
beam=0 i=0 xml status=0/0
beam=0 i=1 xml status=0/0
beam=0 i=2 xml status=0/0
starting beam sizes 3
starting beam sizes 3
beam 0: 0
subbeamCount = 1
ADDING ADDITIONAL HYPOTHESES
REGULAR BEAM EXPANSION
beam 0, subbeam 0: 111
nth->setHypMask
collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 0
collectedCosts[0][0].size() = 0
build new XmlOptionCoveredList for beam 0 hyp 0 cost -0.715873   schlimm:16626 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 1
collectedCosts[0][0].size() = 1
build new XmlOptionCoveredList for beam 0 hyp 0 cost -1.32059    schlecht:2927 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 2
collectedCosts[0][0].size() = 2
build new XmlOptionCoveredList for beam 0 hyp 0 cost -2.13292    übel:34649 ...SUBBEAM 0 COST/KEY
beam 0 hyp 0>0  cost -0.715873   schlimm:16626 ...[0]  ist das
beam 0 hyp 0>1  cost -1.32059    schlecht:2927 ...[0]  ist das
beam 0 hyp 0>2  cost -2.13292    übel:34649 ...[0]  ist das
allotted: 3/3
 toBeRedistributed=0 3/3redistributed: 3/3
merge beam 0 from subbeam 0, hyp 0: schlimm:16626,-0.715873
merge beam 0 from subbeam 0, hyp 1: schlecht:2927,-1.32059
merge beam 0 from subbeam 0, hyp 2: übel:34649,-2.13292
outCosts.size() = 3
outCosts.size() = 3, localBeamSize = 3
beam 0 hyp 0>0  cost -0.715873   schlimm ...[0]  ist das
beam 0 hyp 0>1  cost -1.32059    schlecht ...[0]  ist das
beam 0 hyp 0>2  cost -2.13292    übel ...[0]  ist das
beam 0 hyp 0    cost -0.715873  [0]  schlimm ist das
beam 0 hyp 1    cost -1.32059   [0]  schlecht ist das
beam 0 hyp 2    cost -2.13292   [0]  übel ist das
pruning the beam
histories[i]->Add(beams[i]
histories[i]->Add(beams[i] OK
remaining beam sizes 3
beam 0 hyp 0    cost -0.715873  [0]  schlimm ist das
beam 0 hyp 1    cost -1.32059   [0]  schlecht ist das
beam 0 hyp 2    cost -2.13292   [0]  übel ist das
DONE WITH LOOP, localBeamSize now 3
beam=0 i=0 xml status=0/0
beam=0 i=1 xml status=0/0
beam=0 i=2 xml status=0/0
starting beam sizes 3
starting beam sizes 3
beam 0: 0
subbeamCount = 1
ADDING ADDITIONAL HYPOTHESES
REGULAR BEAM EXPANSION
beam 0, subbeam 0: 111
nth->setHypMask
collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 0
collectedCosts[0][0].size() = 0
build new XmlOptionCoveredList for beam 0 hyp 0 cost -0.849237   </s>:0 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 1
collectedCosts[0][0].size() = 1
build new XmlOptionCoveredList for beam 0 hyp 1 cost -1.38329    </s>:0 ...collectedKeys.size() = 1
collectedKeys[beamNo].size() = 1
collectedKeys[beamNo][subbeam].size() = 2
collectedCosts[0][0].size() = 2
build new XmlOptionCoveredList for beam 0 hyp 2 cost -2.27102    </s>:0 ...SUBBEAM 0 COST/KEY
beam 0 hyp 0>0  cost -0.849237   </s>:0 ...[0]  schlimm ist das
beam 0 hyp 1>1  cost -1.38329    </s>:0 ...[0]  schlecht ist das
beam 0 hyp 2>2  cost -2.27102    </s>:0 ...[0]  übel ist das
allotted: 3/3
 toBeRedistributed=0 3/3redistributed: 3/3
merge beam 0 from subbeam 0, hyp 0: </s>:0,-0.849237
merge beam 0 from subbeam 0, hyp 1: </s>:0,-1.38329
merge beam 0 from subbeam 0, hyp 2: </s>:0,-2.27102
outCosts.size() = 3
outCosts.size() = 3, localBeamSize = 3
beam 0 hyp 0>0  cost -0.849237   </s> ...[0]  schlimm ist das
beam 0 hyp 1>1  cost -1.38329    </s> ...[0]  schlecht ist das
beam 0 hyp 2>2  cost -2.27102    </s> ...[0]  übel ist das
beam 0 hyp 0    cost -0.849237  [0]
beam 0 hyp 1    cost -1.38329   [1]
beam 0 hyp 2    cost -2.27102   [2]
pruning the beam
histories[i]->Add(beams[i]
Add 4 0 -0.849237
Add 4 1 -1.38329
Add 4 2 -2.27102
histories[i]->Add(beams[i] OK
remaining beam sizes 0
DONE WITH LOOP, localBeamSize now 0
[2018-12-07 22:04:15] Best translation 0 : das ist schlimm
[2018-12-07 22:04:15] Total time:  1.234949s wall, 1.100000s user + 0.130000s system = 1.230000s CPU (99.6%)
</pre></code>
</details>

**Comment by phikoehn** [comment]
I uploaded a model that I have been using for testing to http://www.statmt.org/system.tgz.

This also included 10 test sentences, the first on is:
<a translation="Trump">Obama</a> empfängt Net@@ any@@ ahu

Which gives the translation:
Trump welcomes Net@@ any@@ ahu

The model would not mistake these two presidents otherwise.

**Comment by snukky** [comment]
@phikoehn Could you upload vocabs again? The archive contains only symlinks.

I used your command (from translate-ten-xml.sh) and input file (preprocessed with different BPE codes) with http://data.statmt.org/wmt16_systems/de-en/, and the first sentence is translated into:

    Obama welcomes Netanyahu

**Comment by phikoehn** [comment]
I uploaded these to http://www.statmt.org/train.bpe.de.json and http://www.statmt.org/train.bpe.en.json

**Comment by snukky** [comment]
Thanks! I get the proper translation now. It works with other models too, but I needed to set higher `--xml-violation-score` (fixed in the branch "xml-merge") and, more importantly, make sure that words from XML tags occur in the target vocabulary.

**Comment by snukky** [comment]
I've already merged the code with the current master in the branch "xml-merge". I've also fixed the decoding without `--xml-input` (there was an illegal memory access error). I think there is more to do before creating a pull request and putting this into master:

1. Adding regression tests
1. Removing the debug logging
1. General refactorization: including memory allocation, dropping multiple vocabulary loadings, cleaning beam search code, etc.
1. CPU implementation (masking in n-th element)
1. Checking why beam size 1 gives bad translations

I'll add a bunch of regression tests tomorrow and I'm also happy with helping with other things, so I will start some refactorization and cleaning in spare time.

**Comment by dinosaxon** [comment]
Hi all

I would like to ask you if you had the time to work on this feature. I tested on my side and it works fine with my models. However, get OoM errors sometimes.

Thank you

**Comment by snukky** [comment]
Have you tried the "xml-merge" branch? I've fixed some more things, added a bunch of regression tests, and removed bare pointers. The code still needs refactorization, and a higher memory usage can be caused by extensive debugging logs.

**Comment by dinosaxon** [comment]
Yes, I tried this one, about a week ago or so. I will update the code later today or tomorrow and will give it a try.

**Comment by phikoehn** [comment]
I removed all the debugging messages. It is now more than twice as fast.

**Comment by snukky** [comment]
I will rebase with the current master and do more refactoring. If I remember correctly, there were some code added in other classes just to make the debugging possible. I've also left more TODOs for myself when I was doing the first merge with master.

**Comment by phikoehn** [comment]
The debugging cerr messages were all locel - there was only one exception: the loading of the target vocabulary (which got also passed via a function call), which I also removed now.

**Comment by alvations** [comment]
Bumping on issues ;P @phikoehn @snukky @emjotde  @leod @mtresearcher 

Any progress on the xml decoding? Is there anything we can help? Which branch should we look at to contribute to the xml-decoding feature?

Is the https://github.com/ebay-hlt/marian-dev/pull/1 still relevant?

**Comment by mtresearcher** [comment]
I don't think my PR is valid anymore.
